src/OrchardCore/OrchardCore.Media.Abstractions/Indexing/MediaFileIndexingOptions.cs

```diff
@@ -0,0 +1,51 @@
+using System;
+using System.Collections.Generic;
+using System.Linq;
+
+namespace OrchardCore.Media.Indexing
+{
+    public class MediaFileIndexingOptions
+    {
+        private readonly Dictionary<string, Type> _mediaFileTextProviderRegistrations =
+            new Dictionary<string, Type>(StringComparer.OrdinalIgnoreCase);
+
+        public MediaFileIndexingOptions RegisterMediaFileTextProvider<TMediaFileTextProvider>(string fileExtension)
+            where TMediaFileTextProvider : class, IMediaFileTextProvider
+        {
+            return RegisterMediaFileTextProvider(fileExtension, typeof(TMediaFileTextProvider));
+        }
+
+        public MediaFileIndexingOptions RegisterMediaFileTextProvider(string fileExtension, Type providerType)
+        {
+            // Deliberate overwrite behavior so you can override registrations.
```

Comment:
The suggestion here was to add
`var extension = fileExtension.TrimStart('.') + '.'` during registration so that the `.` is always prefixed (this is a lazy singleton, so will only happen once), and it has been normalized.

At which point you can remove the `NormalizeFileExtension` method, because no need to check it? as we know that `Path.GetExtension` includes the `.` right?

Otherwise `NormalizeFileExtension` should prepend the `.` rather than throwing an exception, or it needs a different name, because it's not normalizing ;)

I'd just drop it myself - keep the whole thing simpler

Reply:
This is a published part of the interface that any developer may use, just as `GetRegisteredMediaFileTextProvider()` (though that perhaps less so, I still see the utility in keeping it open). Thus, we need validation. And while as I mentioned having to always include a dot in `AddMediaFileTextProvider()` and `RegisterMediaFileTextProvider()` (something that other developers will use) is inconvenient, perhaps it's acceptable to reduce allocations, as you asked for: Because if we require it in `GetRegisteredMediaFileTextProvider() then it would be confusing to not require the leading dot in these. Thus while `NormalizeFileExtension()` is now indeed incorrectly named (fixed it) if it would prepend the dot then we're back at having additional allocations, or an inconsistent API. I find both of these unfavorable compared to having the current check.